src/wrapper.rs

```diff
@@ -93,102 +93,149 @@ impl DxcOperationResult {
+        unsafe { self.inner.get_result(&mut blob) }.result()?;
+        Ok(DxcBlob::new(blob.unwrap()))
     }
 
     pub fn get_error_buffer(&self) -> Result<DxcBlobEncoding> {
-        let mut blob: ComPtr<IDxcBlobEncoding> = ComPtr::new();
+        let mut blob = None;
 
-        unsafe { self.inner.get_error_buffer(blob.as_mut_ptr()) }.result()?;
-        Ok(DxcBlobEncoding::new(blob))
+        unsafe { self.inner.get_error_buffer(&mut blob) }.result()?;
+        Ok(DxcBlobEncoding::new(blob.unwrap()))
     }
 }
 
 pub trait DxcIncludeHandler {
     fn load_source(&mut self, filename: String) -> Option<String>;
 }
 
-#[repr(C)]
-struct DxcIncludeHandlerWrapperVtbl {
-    query_interface: extern "system" fn(
-        *const com_rs::IUnknown,
-        &com_rs::IID,
-        *mut *mut core::ffi::c_void,
-    ) -> HRESULT,
-    add_ref: extern "system" fn(*const com_rs::IUnknown) -> HRESULT,
-    release: extern "system" fn(*const com_rs::IUnknown) -> HRESULT,
-    #[cfg(not(windows))]
-    complete_object_destructor: extern "system" fn(*const com_rs::IUnknown) -> HRESULT,
-    #[cfg(not(windows))]
-    deleting_destructor: extern "system" fn(*const com_rs::IUnknown) -> HRESULT,
-    load_source: extern "system" fn(*mut com_rs::IUnknown, LPCWSTR, *mut *mut IDxcBlob) -> HRESULT,
-}
+class! {
+    #[no_class_factory]
+    class DxcIncludeHandlerWrapper: IDxcIncludeHandler(IDxcUnknownShim) {
+        // Com-rs intentionally does not support lifetimes in its class structs,
+        // since they live on the stack and their lifetime can be prolonged for
+        // as long as someone keeps a reference through `add_ref()`.
+        // The only way for us to access the library and handler implementation,
+        // which are now intentionally behind a borrow to signify our promise
+        // regarding lifetime, is by transmuting them away and "ensuring" the
+        // class object is discarded at the end of our function call.
 
-#[repr(C)]
-struct DxcIncludeHandlerWrapper<'a, 'i> {
-    vtable: Box<DxcIncludeHandlerWrapperVtbl>,
-    handler: &'i mut dyn DxcIncludeHandler,
-    pinned: Vec<Pin<String>>,
-    library: &'a DxcLibrary,
-}
+        library: &'static DxcLibrary,
+        handler: RefCell<&'static mut dyn DxcIncludeHandler>,
 
-impl<'a, 'i> DxcIncludeHandlerWrapper<'a, 'i> {
-    extern "system" fn query_interface(
-        _me: *const com_rs::IUnknown,
-        _rrid: &com_rs::IID,
-        _ppv_obj: *mut *mut core::ffi::c_void,
-    ) -> HRESULT {
-        HRESULT(0) // dummy impl
+        pinned: RefCell<Vec<Pin<String>>>,
     }
 
-    extern "system" fn dummy(_me: *const com_rs::IUnknown) -> HRESULT {
-        HRESULT(0) // dummy impl
+    impl IDxcUnknownShim for DxcIncludeHandlerWrapper {
+        #[cfg(not(windows))]
+        fn complete_object_destructor(&self) -> HRESULT {
+            HRESULT(0)
+        }
+        #[cfg(not(windows))]
+        fn deleting_destructor(&self) -> HRESULT {
+            HRESULT(0)
+        }
     }
```

Comment:
I thought I fixed this upstream in com-rs (previously cfgs were not at all allowed here) but it turns out on Windows the macro keeps searching for these function pointers.

I'd rather just have https://github.com/microsoft/DirectXShaderCompiler/pull/3793 merged instead of figuring out how to fix this :(

Reply:
https://github.com/microsoft/com-rs/issues/166 Ah, it didn't end up getting fixed.

Reply:
https://github.com/microsoft/com-rs/pull/234.

Reply:
The upstream fix has now been merged, none of this hackery is needed anymore :tada: